File: typescript/src/http.ts
import {
  KairosError,
  AuthError,
  ForbiddenError,
  NotFoundError,
  RateLimitError,
  ValidationError,
  InternalError,
} from './errors';
import type { RateLimitInfo } from './types';

interface HttpClientConfig {
  apiKey: string;
  baseUrl: string;
  timeout: number;
  maxRetries: number;
}

export class HttpClient {
  private apiKey: string;
  private baseUrl: string;
  private timeout: number;
  private maxRetries: number;
  public rateLimit: RateLimitInfo | null = null;

  constructor(config: HttpClientConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.baseUrl;
    this.timeout = config.timeout;
    this.maxRetries = config.maxRetries;
  }

  private updateRateLimitInfo(headers: Headers): void {
    const limitPerMinute = headers.get('X-RateLimit-Limit-Minute');
    const remainingPerMinute = headers.get('X-RateLimit-Remaining-Minute');
    const limitPerHour = headers.get('X-RateLimit-Limit-Hour');
    const remainingPerHour = headers.get('X-RateLimit-Remaining-Hour');
    const resetAt = headers.get('X-RateLimit-Reset');

    if (
      limitPerMinute &&
      remainingPerMinute &&
      limitPerHour &&
      remainingPerHour &&
      resetAt
    ) {
      this.rateLimit = {
        limitPerMinute: parseInt(limitPerMinute, 10),
        remainingPerMinute: parseInt(remainingPerMinute, 10),
        limitPerHour: parseInt(limitPerHour, 10),
        remainingPerHour: parseInt(remainingPerHour, 10),
        resetAt: parseInt(resetAt, 10),
      };
    }
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    this.updateRateLimitInfo(response.headers);

    const contentType = response.headers.get('content-type');
    const isJson = contentType?.includes('application/json');
    const requestId = response.headers.get('X-Request-Id');

    if (!response.ok) {
      let errorData: {
        error?: { code: string; message: string };
      } | null = null;

      if (isJson) {
        try {
          errorData = (await response.json()) as {
            error?: { code: string; message: string };
          } | null;
        } catch {
          // Ignore JSON parse errors
        }
      }

      const errorMessage = errorData?.error?.message || response.statusText;
      const errorCode = errorData?.error?.code || 'UNKNOWN_ERROR';

      switch (response.status) {
        case 400:
          throw new ValidationError(errorMessage, requestId ?? undefined);
        case 401:
          throw new AuthError(errorMessage, requestId ?? undefined);
        case 403:
          throw new ForbiddenError(errorMessage, requestId ?? undefined);
        case 404:
          throw new NotFoundError(errorMessage, requestId ?? undefined);
        case 429: {
          const retryAfter = response.headers.get('Retry-After');
          const retryAfterSeconds = retryAfter ? parseInt(retryAfter, 10) : 60;
          throw new RateLimitError(
            errorMessage,
            retryAfterSeconds,
            requestId ?? undefined,
          );
        }
        case 500:
        case 502:
        case 503:
        case 504:
          throw new InternalError(errorMessage, requestId ?? undefined);
        default:
          throw new KairosError(
            errorCode,
            errorMessage,
            response.status,
            requestId ?? undefined,
          );
      }
    }

    if (isJson) {
      return (await response.json()) as T;
    }

    throw new KairosError(
      'INVALID_RESPONSE',
      'Response is not JSON',
      response.status,
      requestId ?? undefined,
    );
  }

  private async makeRequest<T>(
    method: string,
    path: string,
    body?: unknown,
    retryCount = 0,
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const headers: Record<string, string> = {
        'Authorization': `Bearer ${this.apiKey}`,
        'User-Agent': 'kairos-sdk-js/0.1.0',
      };
      // Only set Content-Type when there is a body — sending it on GET/DELETE
      // causes some Cloudflare Workers to return 404 due to request routing.
      if (body !== undefined) {
        headers['Content-Type'] = 'application/json';
      }

      const response = await fetch(url, {
        method,
        headers,
        body: body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      return await this.handleResponse<T>(response);
    } catch (error) {
      if (error instanceof RateLimitError && retryCount < this.maxRetries) {
        const delayMs = error.retryAfter * 1000;
        await new Promise((resolve) => setTimeout(resolve, delayMs));
        return this.makeRequest<T>(method, path, body, retryCount + 1);
      }

      throw error;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  async get<T>(path: string, params?: Record<string, unknown>): Promise<T> {
    let fullPath = path;

    if (params && Object.keys(params).length > 0) {
      const searchParams = new URLSearchParams();
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined && value !== null) {
          searchParams.append(key, String(value));
        }
      }
      fullPath = `${path}?${searchParams.toString()}`;
    }

    return this.makeRequest<T>('GET', fullPath);
  }

  async post<T>(path: string, body?: unknown): Promise<T> {
    return this.makeRequest<T>('POST', path, body);
  }

  async put<T>(path: string, body?: unknown): Promise<T> {
    return this.makeRequest<T>('PUT', path, body);
  }

  async patch<T>(path: string, body?: unknown): Promise<T> {
    return this.makeRequest<T>('PATCH', path, body);
  }

  async delete<T>(path: string): Promise<T> {
    return this.makeRequest<T>('DELETE', path);
  }
}

File: go/http.go
package kairos

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

// do executes an HTTP request, handles retries on 429, and decodes the response.
func (c *Client) do(ctx context.Context, method, path string, body interface{}, out interface{}) (*RateLimit, error) {
	var reqBody io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("kairos: failed to marshal request body: %w", err)
		}
		reqBody = bytes.NewReader(jsonBody)
	}

	fullURL := c.baseURL + path
	req, err := http.NewRequestWithContext(ctx, method, fullURL, reqBody)
	if err != nil {
		return nil, fmt.Errorf("kairos: failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+c.apiKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "kairos-sdk-go/"+Version)

	var lastErr error
	for attempt := 0; attempt <= c.maxRetries; attempt++ {
		resp, err := c.httpClient.Do(req)
		if err != nil {
			lastErr = fmt.Errorf("kairos: request failed: %w", err)
			continue
		}

		rateLimit := parseRateLimit(resp)

		// Handle 429 Too Many Requests with retry
		if resp.StatusCode == http.StatusTooManyRequests && attempt < c.maxRetries {
			retryAfter := parseRetryAfter(resp)
			select {
			case <-time.After(retryAfter):
				// Retry
				if reqBody != nil {
					// Reset body for retry
					jsonBody, _ := json.Marshal(body)
					req.Body = io.NopCloser(bytes.NewReader(jsonBody))
				}
				resp.Body.Close()
				continue
			case <-ctx.Done():
				resp.Body.Close()
				return rateLimit, ctx.Err()
			}
		}

		respBody, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			return rateLimit, fmt.Errorf("kairos: failed to read response body: %w", err)
		}

		// Handle error responses
		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			var errResp struct {
				Error struct {
					Code      string `json:"code"`
					Message   string `json:"message"`
					RequestID string `json:"request_id"`
				} `json:"error"`
			}
			json.Unmarshal(respBody, &errResp)

			return rateLimit, &KairosError{
				Code:       errResp.Error.Code,
				Message:    errResp.Error.Message,
				StatusCode: resp.StatusCode,
				RequestID:  errResp.Error.RequestID,
			}
		}

		// Decode success response
		if out != nil {
			if err := json.Unmarshal(respBody, out); err != nil {
				return rateLimit, fmt.Errorf("kairos: failed to decode response: %w", err)
			}
		}

		return rateLimit, nil
	}

	return nil, lastErr
}

// get executes a GET request with optional query parameters.
func (c *Client) get(ctx context.Context, path string, query interface{}, out interface{}) (*RateLimit, error) {
	if query != nil {
		queryStr := encodeQueryParams(query)
		if queryStr != "" {
			path = path + "?" + queryStr
		}
	}
	return c.do(ctx, http.MethodGet, path, nil, out)
}

// post executes a POST request.
func (c *Client) post(ctx context.Context, path string, body interface{}, out interface{}) (*RateLimit, error) {
	return c.do(ctx, http.MethodPost, path, body, out)
}

// patch executes a PATCH request.
func (c *Client) patch(ctx context.Context, path string, body interface{}, out interface{}) (*RateLimit, error) {
	return c.do(ctx, http.MethodPatch, path, body, out)
}

// delete executes a DELETE request.
func (c *Client) delete(ctx context.Context, path string, out interface{}) (*RateLimit, error) {
	return c.do(ctx, http.MethodDelete, path, nil, out)
}

// encodeQueryParams converts a struct to URL query parameters.
func encodeQueryParams(v interface{}) string {
	if v == nil {
		return ""
	}

	// Handle ListOptions
	if opts, ok := v.(*ListOptions); ok {
		values := url.Values{}
		if opts.Limit > 0 {
			values.Set("limit", strconv.Itoa(opts.Limit))
		}
		if opts.Offset > 0 {
			values.Set("offset", strconv.Itoa(opts.Offset))
		}
		return values.Encode()
	}

	// Handle ListTasksOptions
	if opts, ok := v.(*ListTasksOptions); ok {
		values := url.Values{}
		if opts.Limit > 0 {
			values.Set("limit", strconv.Itoa(opts.Limit))
		}
		if opts.Offset > 0 {
			values.Set("offset", strconv.Itoa(opts.Offset))
		}
		if opts.Status != "" {
			values.Set("status", opts.Status)
		}
		if opts.Priority != "" {
			values.Set("priority", opts.Priority)
		}
		if opts.AssignedTo != "" {
			values.Set("assigned_to", opts.AssignedTo)
		}
		if opts.GoalID != "" {
			values.Set("goal_id", opts.GoalID)
		}
		if opts.Search != "" {
			values.Set("search", opts.Search)
		}
		return values.Encode()
	}

	return ""
}

// parseRateLimit extracts rate limit info from response headers.
func parseRateLimit(resp *http.Response) *RateLimit {
	rl := &RateLimit{}
	if v := resp.Header.Get("X-RateLimit-Limit-Minute"); v != "" {
		rl.LimitMinute, _ = strconv.Atoi(v)
	}
	if v := resp.Header.Get("X-RateLimit-Remaining-Minute"); v != "" {
		rl.RemainingMinute, _ = strconv.Atoi(v)
	}
	if v := resp.Header.Get("X-RateLimit-Limit-Hour"); v != "" {
		rl.LimitHour, _ = strconv.Atoi(v)
	}
	if v := resp.Header.Get("X-RateLimit-Remaining-Hour"); v != "" {
		rl.RemainingHour, _ = strconv.Atoi(v)
	}
	if v := resp.Header.Get("X-RateLimit-Reset"); v != "" {
		rl.Reset, _ = strconv.ParseInt(v, 10, 64)
	}
	return rl
}

// parseRetryAfter extracts the Retry-After header.
func parseRetryAfter(resp *http.Response) time.Duration {
	if retryAfter := resp.Header.Get("Retry-After"); retryAfter != "" {
		if seconds, err := strconv.Atoi(retryAfter); err == nil {
			return time.Duration(seconds) * time.Second
		}
	}
	return time.Second // Default to 1 second
}
